### PR TITLE
CI: Pick a M365 app from a pool

### PR DESCRIPTION
**Pull request by vkamra**
## Description

Attempt to reduce flakiness in CI due to a single CI app accessing the same test users mailbox across different CI jobs.

Introduces logic to pick one of the 4 pre-configured apps for CI.

## Does this PR need a docs update or release note?

- [ ] :white_check_mark: Yes, it's included
- [ ] :clock1: Yes, but in a later PR
- [x] :no_entry: No 

## Type of change

- [ ] :sunflower: Feature
- [ ] :bug: Bugfix
- [ ] :world_map: Documentation
- [ ] :robot: Test
- [x] :computer: CI/Deployment
- [ ] :hamster: Trivial/Minor

## Issue(s)

* #1776 

## Test Plan

- [ ] :muscle: Manual
- [x] :zap: Unit test
- [ ] :green_heart: E2E

### CODE REVIEW

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -129,13 +164,18 @@ jobs:
   # ----------------------------------------------------------------------------------------------------
 
   Test-Suite-Trusted:
-    needs: [Precheck, Checkout]
+    needs: [Precheck, Checkout, SetM365App]
```
[the remark this answers is not in the document]
Comment (vkamra) [reply .github/workflows/ci.yml:167]: Yes - will do that as a separate PR.

File: .github/workflows/ci.yml [hunk 2]
```diff
@@ -50,6 +50,41 @@ jobs:
...
+      - name: Figure out which client id to use
+        id: roundrobin
+        run: |
+          slot=$((GITHUB_RUN_NUMBER % 4))
+          echo "CLIENT_APP_SLOT=$slot" >> $GITHUB_OUTPUT
+
+          case $slot in
+
+            0)
+              echo "CLIENT_ID_ENV=CLIENT_ID" >> $GITHUB_OUTPUT
```
[the remark this answers is not in the document]
Comment (vkamra) [reply .github/workflows/ci.yml:71]: No - this is complete. Anything we should change here?

File: .github/workflows/ci.yml [hunk 3]
```diff
@@ -50,6 +50,41 @@ jobs:
...
+    runs-on: ubuntu-latest
+    outputs:
+      client_app_slot: ${{ steps.roundrobin.outputs.CLIENT_APP_SLOT }}
+      client_id_env: ${{ steps.roundrobin.outputs.CLIENT_ID_ENV }}
+      client_secret_env: ${{ steps.roundrobin.outputs.CLIENT_SECRET_ENV }}
+    steps:
+      - name: Figure out which client id to use
+        id: roundrobin
+        run: |
+          slot=$((GITHUB_RUN_NUMBER % 4))
```
Comment (ryanfkeepers) [inline .github/workflows/ci.yml:65]: per docs on run_number:
```
A unique number for each run of a particular workflow in a repository. This number begins at 1 for the workflow's first run, and increments with each new run. This number does not change if you re-run the workflow run. For example, 3.
```
If I'm reading that right, it sounds like runs will always kick off on the same number, leading to re-use of the same client id for each new PR.  

We might get something more random with `GITHUB_RUN_ID`
```
A unique number for each workflow run within a repository. This number does not change if you re-run the workflow run. For example, 1658821493.
```

**Review by ryanfkeepers — APPROVED** [review]

File: .github/workflows/ci.yml (hunk 3 above)
Comment (vkamra) [reply .github/workflows/ci.yml:65]: I did look at that. My thought was that we actually don't want something random but rather monotonically increasing would give us the behavior we want.

**Pull request merged**